ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAutoML.scala

```diff
@@ -0,0 +1,386 @@
+import org.apache.spark.sql.{Dataset, SQLContext}
+import org.json4s.JsonAST.JArray
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.{JNull, JString, JValue}
+import water.Key
+import water.fvec.{Frame, H2OFrame}
+import water.support.{H2OFrameSupport, ModelSerializationSupport}
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+
+/**
+  * Base class for H2O algorithm wrapper as a Spark transformer.
+  */
+class H2OAutoML(val automlBuildSpec: Option[AutoMLBuildSpec], override val uid: String)
+               (implicit hc: H2OContext, sqlContext: SQLContext)
+  extends Estimator[H2OMOJOModel] with MLWritable with H2OAutoMLParams {
+
+  def this()(implicit hc: H2OContext, sqlContext: SQLContext) = this(None, Identifiable.randomUID("automl"))
+
+  def this(uid: String, hc: H2OContext, sqlContext: SQLContext) = this(None, uid)(hc, sqlContext)
+
+  override def fit(dataset: Dataset[_]): H2OMOJOModel = {
+    val spec = automlBuildSpec.getOrElse(new AutoMLBuildSpec)
+
+    // override the buildSpec with the configuration specified directly on the estimator
+
+    val input = hc.asH2OFrame(dataset.toDF())
+    // check if we need to do any splitting
+    if (getRatio() < 1.0) {
+      // need to do splitting
+      val keys = split(input, hc)
+      spec.input_spec.training_frame = keys(0)
+      if (keys.length > 1) {
+        spec.input_spec.validation_frame = keys(1)
+      }
+    } else {
+      spec.input_spec.training_frame = input._key
+    }
+
+
+    val trainFrame = spec.input_spec.training_frame.get()
+    if (getAllStringColumnsToCategorical()) {
+      H2OFrameSupport.allStringVecToCategorical(trainFrame)
+    }
+    spec.input_spec.response_column = getPredictionsCol()
+    spec.input_spec.fold_column = getFoldColumn()
+    spec.input_spec.weights_column = getWeightsColumn()
+    spec.input_spec.ignored_columns = getIgnoredColumns()
+    spec.feature_engineering.try_mutations = getTryMutations()
+    spec.build_models.exclude_algos = getExcludeAlgos()
+    spec.build_control.project_name = getProjectName()
+    spec.build_control.loss = getLoss()
+    spec.build_control.stopping_criteria.set_max_runtime_secs(getMaxRuntimeSecs())
+    spec.build_control.stopping_criteria.set_stopping_rounds(getStoppingRounds())
+    spec.build_control.stopping_criteria.set_stopping_tolerance(getStoppingTolerance())
+    spec.build_control.stopping_criteria.set_stopping_metric(getStoppingMetric())
+    spec.build_control.nfolds = getNfolds()
+
+    water.DKV.put(trainFrame)
+    val aml = new AutoML(Key.make(uid), new Date(), spec)
+    AutoML.startAutoML(aml)
+    // Block until AutoML finishes
+    aml.get()
+    val model = new H2OMOJOModel(ModelSerializationSupport.getMojoData(aml.leader()))
+    model.setConvertUnknownCategoricalLevelsToNa(true)
+    model
+  }
+
+  private def split(fr: H2OFrame, hc: H2OContext): Array[Key[Frame]] = {
+    val trainKey = Key.make[Frame]("train")
```

Comment:
What if `train` key already exist?

Reply:
Good, point, thanks. I will fix this before the release